src/peergos/shared/user/fs/cryptree/CryptreeNode.java

```diff
@@ -7,104 +7,785 @@
+                                                            SymmetricKey parentparentKey,
+                                                            byte[] newMapKey,
+                                                            NetworkAccess network,
+                                                            SafeRandom random,
+                                                            Hasher hasher) {
+        if (! isDirectory) {
+            throw new IllegalStateException("Copy to only valid for directories!");
+        }
+        SymmetricKey newWriteBaseKey = SymmetricKey.random();
+        SymmetricKey parentKey = getParentKey(us.rBaseKey);
+        FileProperties props = getProperties(parentKey);
+        Optional<SigningPrivateKeyAndPublicHash> newSigner = Optional.empty();
+        RelativeCapability nextChunk = new RelativeCapability(Optional.empty(), random.randomBytes(32), newReadBaseKey, Optional.empty());
+        RelativeCapability parentLink = new RelativeCapability(Optional.empty(), newParentCap.getMapKey(), parentparentKey, Optional.empty());
+        DirAndChildren dirWithLinked = CryptreeNode.createDir(MaybeMultihash.empty(), newReadBaseKey, newWriteBaseKey, newSigner, props,
+                Optional.of(parentLink), parentKey, nextChunk, hasher);
+        CryptreeNode da = dirWithLinked.dir;
+        SymmetricKey ourNewParentKey = da.getParentKey(newReadBaseKey);
+        WritableAbsoluteCapability ourNewCap = new WritableAbsoluteCapability(newParentCap.owner, newParentCap.writer, newMapKey, newReadBaseKey, newWriteBaseKey);
+
+        return this.getChildren(network, us).thenCompose(RFPs -> {
+            // upload new metadata blob for each child and re-add child
+            CompletableFuture<CryptreeNode> reduce = RFPs.stream().reduce(CompletableFuture.completedFuture(da), (dirFuture, rfp) -> {
+                SymmetricKey newChildReadKey = rfp.fileAccess.isDirectory() ? SymmetricKey.random() : rfp.capability.rBaseKey;
+                SymmetricKey newChildWriteKey = SymmetricKey.random();
+                byte[] newChildMapKey = random.randomBytes(32);
+                WritableAbsoluteCapability newChildCap = new WritableAbsoluteCapability(ourNewCap.owner,
+                        ourNewCap.writer, newChildMapKey, newChildReadKey, newChildWriteKey);
+                return rfp.fileAccess.copyTo(rfp.capability, newChildReadKey,
+                        ourNewCap, newEntryWriter, ourNewParentKey, newChildMapKey, network, random, hasher)
+                        .thenCompose(newChildFileAccess -> {
+                            return dirFuture.thenCompose(dirAccess ->
+                                    dirAccess.addChildAndCommit(ourNewCap.relativise(newChildCap), ourNewCap,
+                                            newEntryWriter, network, random, hasher));
+                        });
+            }, (a, b) -> a.thenCompose(x -> b)); // TODO Think about this combiner function
```

Comment:
TODO msg

Reply:
Yep that's been there for a while. It's still a todo :-)